### PR TITLE
make `fold.expresion.cpp` work with negative values

### LINKED ISSUE #267: Chapter 2 exercise answer bug
The exercise answer could not handle negative values:

```cpp
template<typename ... T>
auto average(T ... t) {
    return (t + ... ) / sizeof...(t);
}
int main() {
    std::cout << average(1, 2, 3, 4, 5, -6, -7, -8, -9, -10) << std::endl;
}
```

It would return `1844674407370955159` on my side since it deduced the return value to be unsigned int.

### PR DESCRIPTION
**Pull request by Delta456**
resolve #267 

## Description

Please describe the motivation of this pull request, and what problem was solved in this PR.

## Change List

- Resolve the error content of https://github.com/changkun/modern-cpp-tutorial/issues/267

## Reference

I added a `static_cast<double>` to make it work with negative values.

---

## 说明

此处详细说明 PR 的动机是什么、解决了什么样的问题。

## 变化箱单

- 修复了 XXX 的 typo 错误
- 增加了 XXX 相关的说明
- 解决了关于 XXX 的描述性错误

## 参考文献

如果有请注明

### CODE REVIEW

**Comment by Mayday646** [comment]
Do not use 'auto' for the return of template functions, but use 'double' instead

File: exercises/2/fold.expresion.cpp [hunk 1]
```diff
@@ -11,8 +11,8 @@
 #include <iostream>
 template<typename ... T>
 auto average(T ... t) {
-    return (t + ... ) / sizeof...(t);
+    return static_cast<double>((t + ... )) / sizeof...(t);
```
Comment (frederick-vs-ja) [inline exercises/2/fold.expresion.cpp:14]: 我个人建议是模仿 `<cmath>` 中浮点函数的额外重载，把整数类型转换为 `double` 而浮点类型不变。这样可以避免 `long double` 被转换为 `double`。

```suggestion
    constexpr auto integer_to_double = [](auto x) {
         if constexpr (std::is_integral_v<decltype(x)>)
             return static_cast<double>(x);
         else
             return x;
    };
    return (integer_to_double(t) + ... ) / sizeof...(t);
```

**Comment by Devanshu-0808** [comment]
static_caste<double>() it will explicitly change the data type of t to double
``` c++
#include <iostream>

template<typename ... T>
auto average(T ... t) {
    return (static_cast<double>(t) + ...) / sizeof...(t);
}

int main() {
    std::cout << average(1, 2, 3, 4, 5, -6, -7, -8, -9, -10) << std::endl;
}
```

File: exercises/2/fold.expresion.cpp (hunk 1 above)
Comment (Devanshu-0808) [reply exercises/2/fold.expresion.cpp:14]: static_caste() it will explicitly change the data type of t to double
``` c++
#include <iostream>

template<typename ... T>
auto average(T ... t) {
    return (static_cast<double>(t) + ...) / sizeof...(t);
}

int main() {
    std::cout << average(1, 2, 3, 4, 5, -6, -7, -8, -9, -10) << std::endl;
}
}
```

**Review by Devanshu-0808 — CHANGES_REQUESTED** [review]

**Comment by MosZhang7** [comment]
English: The large amount of defensive code added to handle various forms of incidental complexity—such as precision conversions, exceptional cases, and boundary conditions—unnecessarily inflates a function that is originally intended to demonstrate the core idea. This deviates from its purpose as an exercise function designed to test a specific concept. I do not recommend expanding the function excessively; it should remain concise, easy to understand, and focused on the basic scenario. If needed, they can be limited to what is discussed in this chapter, or we may simply keep the first revision, which only prevents unintended integer promotion behavior. For example, although this chapter mentions that a variadic template parameter pack may be empty and lead to a division-by-zero crash, such edge handling is beyond the scope of this exercise. Therefore, I suggest modifying it as follows:

Chinese: 为处理各种附带复杂性如精度转换, 异常情况, 边界条件等而加入的大量防御性代码, 使得原本用于展示核心思路的函数不必要的膨胀. 这脱离了它原本作为课后习题应该用来检验知识点的目的. 我不建议函数过多膨胀, 应该简洁易懂, 考虑基本情况即可.  如果要考虑,可以考虑本章提到的,或者保持第一版的修改,仅仅防止integer promotion. 例如本章提到的: 可变模板参数, 可以不传参数, 但会导致除零崩溃。因此建议修改如下:
``` C++
#include <iostream>

template <typename... T>
auto average(T... t) {
  return static_cast<double>((0 + ... + t)) /
         (sizeof...(t) > 0 ? sizeof...(t) : 1);
}

int main() {
  std::cout << average(1, 2, 3, 4, 5, -6, -7, -8, -9, -10) << std::endl;
  std::cout << average() << std::endl;
}
```
